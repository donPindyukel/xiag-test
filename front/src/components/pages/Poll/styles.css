.content-wrap {
  padding: 20px 40px;
  .poll-form-wrap {
    margin: 40px auto;
    max-width: 505px;
    display: flex;
    flex-direction: column;

    .title-wrap {
      h1 {
        font-size: 25px;
        font-weight: normal;
        margin: 0;
        padding: 0;
        line-height: 1.5;
        color: var(--fontColor);
      }
    }

    .poll-wrap {
      margin: 20px 0 40px;
      padding: 20px 0;
      border-top: 1px #eee solid;
      border-bottom: 1px #eee solid;

      .name-input-wrap {
        width: 100%;
        margin-bottom: 20px;

        .name-input-title {
          font-size: 16px;
          margin-bottom: 10px;
          line-height: 1.5;
          color: var(--fontColor);
        }

        input {
          width: 100%;
          box-sizing: border-box;
        }
      }
    }

    .answers-variants {
      display: flex;
      justify-content: flex-start;

      .answer-item {
        margin-bottom: 10px;
        margin-right: 20px;

        label {
          font-size: 15px;
          line-height: 1.5;
          color: var(--fontColor);
        }

        input {
          font-size: 16px;
          box-sizing: border-box;
          margin: 3px 3px 0px 5px;
        }
      }
    }

    .submit-button-wrap {

      button {
        padding: 0.6rem 2.5rem;
        font-size: 16px;
        margin: 20px 0 0;
        font-family: "Arial", sans-serif;

        &:hower {
          box-shadow: none;
        }

        &:disabled {
          background: var(--borderColor);
          box-shadow: none;
        }
      }
    }
  }
}
