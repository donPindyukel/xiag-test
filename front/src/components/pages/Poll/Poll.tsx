import React from "react";
import { API } from "Services/api";
import { storageService } from "Services/storageService";
import { RouteComponentProps } from "react-router";
import {
  IAnswer,
  IAnswerRes,
  IRoutePollParams,
  IStatData,
  IUserAnswerRequest
} from "Models";
import { Results } from "Elements";
import { mapService } from "Services/mappingService";
import { SocketHandler } from "Services/socketService";
import "./styles.css";

export interface IPollState {
  question: string;
  answers: IAnswer[];
  name: string;
  answer: string;
  isFormValid: boolean;
  isUserAnswered: boolean;
  statData: IStatData[];
}

export type IPollProps = RouteComponentProps<IRoutePollParams>;

export class Poll extends React.Component<IPollProps, IPollState> {
  state = {
    question: "",
    answers: [],
    name: "",
    answer: "",
    isFormValid: false,
    isUserAnswered: false,
    statData: []
  };

  socket: any = null;

  componentDidMount() {
    this.checkUserData();
    this.getPollData();
    this.getResultData();
    this.getConnection();
  }

  async getConnection() {
    this.socket = new SocketHandler(
      this.statDataUpdate,
      this.props.match.params.pollId
    );
  }

  async getResultData() {
    try {
      const pollResultsResp = await API.getPollResults();
      const resultData = mapService.pollResultsMapper(
        pollResultsResp.data,
        this.props.match.params.pollId
      );
      this.setState({
        statData: [...resultData]
      });
    } catch (err) {
      console.log("Error occurred", err);
    }
  }

  async getPollData() {
    try {
      const response = await API.getPoll(this.props.match.params.pollId);
      if (!response.data) {
        this.goBack();
      } else {
        const question = response.data.question;
        const answers = response.data.answers.map((answer: IAnswerRes) => ({
          title: answer.title,
          answerId: answer._id
        }));
        this.setState({
          question,
          answers
        });
      }
    } catch (err) {
      this.goBack();
    }
  }

  async checkUserData() {
    try {
      const userId = storageService.getUserId();
      if (userId) {
        const userResponse = await API.getUser(userId);
        if (
          storageService.checkUser(
            userResponse.data,
            this.props.match.params.pollId
          )
        ) {
          this.setState({
            isUserAnswered: true
          });
        }
        const name = userResponse.data.name;
        this.setState({
          name
        });
      }
    } catch (err) {
      this.goBack();
    }
  }

  statDataUpdate = () => {
    this.getResultData();
  };

  inputNameHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    const name = e.target.value;
    this.setState(
      {
        name
      },
      this.formValidation
    );
  };

  selectAnswerChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const answer = e.target.value;
    this.setState(
      {
        answer
      },
      this.formValidation
    );
  };

  submitClickHandler = async () => {
    try {
      let request: any;
      let id: string;
      const userId = storageService.getUserId();
      if (!userId) {
        request = await API.createUser(this.state.name);
        storageService.putLocalStorage(request.data._id);
        id = request.data._id;
      } else {
        id = userId;
      }
      const params: IUserAnswerRequest = {
        id,
        answer: {
          questionId: this.props.match.params.pollId,
          answerId: this.state.answer
        }
      };
      const sendAnswerRes = await API.sendAnswer(params);
      this.setState({
        isUserAnswered: true
      });
      this.socket.addVote();
      this.statDataUpdate();
    } catch (err) {
      console.log("Error occurred", err);
    }
  };

  formValidation = () => {
    const { name, answer } = this.state;
    if (name && answer) {
      this.setState({
        isFormValid: true
      });
    } else {
      this.setState({
        isFormValid: false
      });
    }
  };

  goBack() {
    this.props.history.push({
      pathname: `/`
    });
  }

  answerRender = (answer: IAnswer) => {
    return (
      <div className="answer-item" key={answer.answerId}>
        <label htmlFor={`item-${answer.answerId}`} className="answer-item">
          <input
            id={`item-${answer.answerId}`}
            name="answer"
            type="radio"
            onChange={this.selectAnswerChange}
            value={answer.answerId}
          />
          {answer.title}
        </label>
      </div>
    );
  };

  questionFormRender() {
    const { question, answers, isFormValid, name, isUserAnswered } = this.state;
    const userId = storageService.getUserId();
    return (
      <div className="poll-form-wrap">
        <div className="title-wrap">
          <h1>{question}</h1>
        </div>
        {!isUserAnswered && (
          <div className="poll-wrap">
            <div className="name-input-wrap">
              <div className="name-input-title">Your name:</div>
              <input
                type="text"
                disabled={!!userId}
                value={name}
                onChange={this.inputNameHandler}
              />
            </div>
            <div className="answers-variants">
              {answers.map(this.answerRender)}
            </div>
            <div className="submit-button-wrap">
              <button disabled={!isFormValid} onClick={this.submitClickHandler}>
                Submit
              </button>
            </div>
          </div>
        )}
        <div className="results-wrap">
          <Results answers={answers} statData={this.state.statData} />
        </div>
      </div>
    );
  }

  render() {
    const { question, answers } = this.state;
    if (question && answers.length) {
      return <div className="content-wrap">{this.questionFormRender()}</div>;
    } else {
      return <div>Loading...</div>;
    }
  }

  componentWillUnmount() {
    this.socket.disconnection();
  }
}
