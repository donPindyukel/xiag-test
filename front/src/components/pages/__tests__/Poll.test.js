import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { Poll } from "../Poll/Poll";

describe("<Poll />", () => {
  let component;

  describe("Poll", () => {
    beforeAll(() => {
      component = shallow(<Poll />);
    });

    it("renders the component", () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});