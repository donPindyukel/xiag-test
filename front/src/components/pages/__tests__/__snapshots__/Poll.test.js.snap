// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Poll /> Poll renders the component 1`] = `
<div>
  Loading...
</div>
`;
